rfd/0169-auto-updates-linux-agents.md

```diff
@@ -0,0 +1,382 @@
+  proxy: mytenant.teleport.sh
+  # enabled specifies whether auto-updates are enabled, i.e., whether teleport-updater update is allowed to update the agent.
+  enabled: true
+  # active_version specifies the active (symlinked) deployment of the telepport agent.
+  active_version: 15.1.1
+```
+
+backup.yaml:
+```
+version: v1
+kind: config_backup
+spec:
+  # proxy address from the backup
+  proxy: mytenant.teleport.sh
+  # version from the backup
+  version: 15.1.0
+  # time the backup was created
+  creation_time: 2020-12-09T16:09:53+00:00
+```
+
+### Runtime
+
+The agent-updater will run as a periodically executing systemd service which runs every 10 minutes.
+The systemd service will run:
+```shell
+$ teleport-updater update
+```
+
+After it is installed, the `update` subcommand will no-op when executed until configured with the `teleport-updater` command:
+```shell
+$ teleport-updater enable --proxy mytenant.teleport.sh
+```
+
+If the proxy address is not provided with `--proxy`, the current proxy address from `teleport.yaml` is used.
+
+On servers without Teleport installed already, the `enable` subcommand will change the behavior of `teleport-update update` to update teleport and restart the existing agent, if running.
+It will also run update teleport immediately, to ensure that subsequent executions succeed.
+
+The `enable` subcommand will:
+1. Configure `updates.yaml` with the current proxy address and set `enabled` to true.
+2. Query the `/v1/webapi/ping` endpoint.
+3. If the current updater-managed version of Teleport is the latest, and teleport package is not installed, quit.
+4. If the current updater-managed version of Teleport is the latest, but the teleport package is installed, jump to (14).
+5. Ensure there is enough free disk space to upgrade Teleport.
+6. Download the desired Teleport tarball specified by `agent_version` and `server_edition`.
+7. Download and verify the checksum (tarball URL suffixed with `.sha256`).
+8. Extract the tarball to `/var/lib/teleport/versions/VERSION`.
+9. Replace any existing binaries or symlinks with symlinks to the current version.
+10. Backup /var/lib/teleport into `/var/lib/teleport/versions/OLD-VERSION/backup/teleport`
+11. Restart the agent if the systemd service is already enabled.
+12. Set `active_version` in `updates.yaml` if successful or not enabled.
+13. Replace the symlinks/binaries and `/var/lib/teleport` and quit (exit 1) if unsuccessful.
+14. Remove any `teleport` package if installed.
+15. Verify the symlinks to the active version still exists.
+16. Remove all stored versions of the agent except the current version.
+
+The `disable` subcommand will:
+1. Configure `updates.yaml` to set `enabled` to false.
+
+When `update` subcommand is otherwise executed, it will:
+1. Check `updates.yaml`, and quit (exit 0) if `enabled` is false, or quit (exit 1) if `enabled` is true and no proxy address is set.
```

Comment:
The step 1 should be creating a lock somewhere with a socket to avoid two update instances to run concurrently. Having two update process running simultaneously would result in non-deterministic behaviour.

Reply:
Added a lock